iio: adc: ad7292: add support for reading single conversion results

marceloschmitt1 opened the pull request:
Add a channel for configuring the ADC sequence register.
Change interal spi buffer to 4 byte size for being able issue 4 byte
write commands.
Add a spi_reg_write function to on AD7292 internal registers.
Add a new case to read_raw to allow reading from ADC sequence register.
Refactor spi_reg_read functions to work properly with the new spi buffer
size.

Signed-off-by: Marcelo Schmitt <marcelo.schmitt1@gmail.com>

commodo commented on drivers/iio/adc/ad7292.c:
```diff
@@ -27,6 +34,13 @@
+# define AD7292_ADC_SEQ_CHAN						\
+{									\
+	.type = IIO_VOLTAGE,						\
+	.info_mask_separate = BIT(IIO_CHAN_INFO_ENABLE),		\
```
the sequencer enable/disable is not an IIO channel;
this mechanism is useful when you want to read multiple channels in a single read operation;
this can usually be coupled with IIO buffer;

I would suggest maybe enabling the sequencer by default, and configure it to read from the channels that are enabled.
This would mean that if you want to read a single channel at a fast rate, it would probably be slower than usual, but that's a case we can think of fixing later.
An example [that comes to mind] is AD7606.
But definitely check for other drivers that implement sequencer reading.

marceloschmitt1 replied:
I added DT nodes to sepecify which ADC channel should be enabled by default in the sequencer.

I found other three drivers that apparently have some kind ADC of sequencer:
- AD7949 (iio tree)
- xilinx-ams (rpi tree)
- stm32-adc (both rpi and iio tree)

I'll try to add a triggered buffer and associated interrup to implement sequenced reading in AD7292 looking at these four other driver examples.

marceloschmitt1 commented:
### Changelog V2:

- Added DT nodes to specify what ADC channels should be active
- Updated DT binding docs
- Removed the ADC sequencer channel
- Cleared `write_raw` function
- Added a function to set up sequencer register according to DT entries

marceloschmitt1 commented:
### Changelog V3:
- removed DT stuff related to ADC sequencer initialization
- added defines for the ADC conversion command and ADC conversion result registers
- added a `conv_command` function to issue conversion commands to channels
- updated `read_raw` to issue conversion commands to ADC channels

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -103,10 +128,38 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
+			*val = ad7292_conv_command(st, AD7292_REG_ADC_CH7);
+			break;
+		default:
+			return -EINVAL;
```
This method adds a lot of code duplication. Another way that you can do it is by placing the channel register addresses in an array, something like this:
```
static unsigned in ad7292_channel_address[8] = {
       AD7292_REG_ADC_CH0,  AD7292_REG_ADC_CH1, AD7292_REG_ADC_CH2, AD7292_REG_ADC_CH3,
       AD7292_REG_ADC_CH4,  AD7292_REG_ADC_CH5, AD7292_REG_ADC_CH6, AD7292_REG_ADC_CH7,
}
```
```
...
unsigned int ch_addr;
...
case IIO_CHAN_INFO_RAW:
       ch_addr = ad7292_channel_address[chan->channel]
       *val = ad7292_conv_command(st, ch_addr);
       break;
```

marceloschmitt1 replied:
OK, updated accordingly. Thanks.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -16,6 +16,16 @@
 
```
The commit message should be more descriptive of what is actually implemented. I would change it to something like:
iio: adc: ad7292: Add support for reading single conversion results

marceloschmitt1 replied:
I changed the commit message and rewrite part of the description in the hope of making it a little bit clearer. Thanks.

marceloschmitt1 replied:
There are some settings in the configuration register bank that may be changed to customize ADC readings on AD7292. If not by IIO channels, how would be the best way of configuring those subregisters?

stefpopa replied:
Indeed, we need to find the best way to deal with different configurations. Generally speaking, we  have two options:
* If the configuration defines the hardware, then we can define it in the device tree. For instance, for our application we might need to use some GPIOs. We can define them in the device tree, and then at probe, we can set the `Digital output driver` or `Digital I/O function `subregisters. 
* Another option, used for configurations that might change after probe, are the iio device attributes. Preferably, it is great if we can use some of the attributes that are already defined, such as `sampling_frequency`. These attributes are documented here: https://github.com/analogdevicesinc/linux/blob/master/Documentation/ABI/testing/sysfs-bus-iio. However, sometimes we need to define our own custom attributes, which is fine as long as they are documented. Here's an example of a custom attribute: https://github.com/analogdevicesinc/linux/blob/master/Documentation/ABI/testing/sysfs-bus-iio-frequency-adf4371#L54

marceloschmitt1 replied:
Thanks. I believe the second option is more suitable for this case. Going to try to implement attributes for these settings.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -94,6 +104,21 @@ static int ad7292_spi_subreg_read(struct ad7292_state *st, unsigned int addr,
 	return (be16_to_cpu(st->data.d16) >> shift);
 }
 
+static int ad7292_conv_command(struct ad7292_state *st, unsigned int chan_addr)
```
I would rename this function to something like `ad7292_single_conversion()`

marceloschmitt1 replied:
OK. Updated to `ad7292_single_conversion`.

marceloschmitt1 commented:
### Changelog V4:
- added an array to group the ADC conversion register's address and access them by id
- renamed `ad7292_conv_command` to `ad7292_single_conversion`
- shortened `read_raw` using the newly added conversion register address array

marceloschmitt1 commented:
### Changelog V5:
- renamed patch title to fit int 50 characters
- fixed ADC reading to return only the measurement bits from the ADC conversion result register
- added a patch to return each ADC channel scale according to AD7292 internal settings

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -103,10 +140,13 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
-		*val = 1000;
-
+		ch_addr = ad7292_channel_address[chan->channel];
+		*val = AD7292_ADC_RSLT(ad7292_single_conversion(st, ch_addr));
```
It is always nicer to store the result of the conversion in a local variable and then use the macro, it makes the code easier to read. Also, your `ad7292_single_conversion()` function might return negative values in case of error, so this needs to be checked as well. Something like this:
```
int ret;
ret = ad7292_single_conversion(st, ch_addr);
if (ret < 0)
       return ret;

*val = AD7292_ADC_RSLT(ret);

return IIO_VAL_INT;
```

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -16,9 +17,23 @@
 
+/* AD7292_REG_ADC_CONVERSION */
+#define AD7292_ADC_RSLT_MASK		GENMASK(15, 6)
+#define AD7292_ADC_RSLT(x)		FIELD_GET(AD7292_ADC_RSLT_MASK, x)
```
nitpick: I would rename this macro to `AD7292_ADC_DATA`

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -37,7 +43,9 @@
-	.info_mask_separate = BIT(IIO_CHAN_INFO_RAW),			\
+	.info_mask_separate =						\
+	BIT(IIO_CHAN_INFO_RAW) |					\
+	BIT(IIO_CHAN_INFO_SCALE),					\
```
indentation here looks strange, I would do it like this:
```
	.info_mask_separate = BIT(IIO_CHAN_INFO_RAW)  | \
	                      BIT(IIO_CHAN_INFO_SCALE), \
```

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -142,16 +150,54 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
+						AD7292_BANK_REG_VIN_RNG0, 2);
+
+		range1 = ad7292_spi_subreg_read(st, AD7292_REG_CONF_BANK,
+						AD7292_BANK_REG_VIN_RNG1, 2);
```
`ad7292_spi_subreg_read()` function might return negative values in case of error. You need to check the return value before you proceed.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -142,16 +150,54 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
+		range1 = ad7292_spi_subreg_read(st, AD7292_REG_CONF_BANK,
+						AD7292_BANK_REG_VIN_RNG1, 2);
+
+		if (samp_mode & BIT((chan->channel + 8))) {
```
You should define a mask and use it here instead of creating one on the go

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -142,16 +150,54 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
+
+			*val *= 4;
+		}
+		*val2 = 10;
```
This whole way of calculating the scale does not look straightforward, so it would be useful to add a comment with a formula which describes in more detail what is happening.

marceloschmitt1 replied:
OK, thanks. Tackled most suggestions except the BUSY one for now.

stefpopa commented on drivers/iio/adc/ad7292.c:
```diff
@@ -142,16 +150,54 @@ static int ad7292_read_raw(struct iio_dev *indio_dev,
 	struct ad7292_state *st = iio_priv(indio_dev);
 	unsigned int ch_addr;
+	int samp_mode;
+	int range0, range1;
```
nitpick: you can probably fit all these `int `variables on one line.

dbogdan commented on drivers/iio/adc/ad7292.c:
```diff
@@ -94,6 +123,22 @@ static int ad7292_spi_subreg_read(struct ad7292_state *st, unsigned int addr,
+	st->data.d8[0] = chan_addr;
+	st->data.d8[1] = AD7292_RD_FLAG_MSK(AD7292_REG_CONV_COMM);
+
+	ret = spi_write_then_read(st->spi, &st->data.d8, 2, &st->data.d16, 2);
```
Marcelo, I don't think that this implementation will work. After the start conversion command is sent, you should wait until the conversion is actually done (BUSY signal goes low). Maybe we can use two spi_sync() calls (one for the start command and another for reading the data) instead of spi_write_then_read(). Using spi_sync(), you gain control over the CS signal (have a look at cs_change member of spi_transfer struct).

marceloschmitt1 replied:
Hi Dragos, you're right. According to datasheet page 6, the time needed to complete an ADC conversion may vary from 900 nano (typical) to a max of 5.85 microseconds. I'm not sure of how much time it takes between the finish of a write and the beginning of a read that occur within a `spi_write_then_read` call but, if get more than some microseconds, the reading might indeed get corrupted. I was wondering what would be the best way to tackle it: using `spi_sync()` with a safe delay after issuing conversion commands, or bounding an IRQ to the BUSY pin to trigger the reading when ADC conversion gets finished. Testing the reading on the evaluation board gave me at first a feeling of right readings in IIO direct mode, however, I suspect I might not get the same feeling if get the AD7292 to do ADC conversion continuously in a buffer triggered mode. =/
Any other comment?

dbogdan replied:
In the first version, I would only add a delay. Later, we can add support for the BUSY signal too.

marceloschmitt1 replied:
Hi, I'm still alive =p. Sorry for the 3 week disappear. I replaced the `spi_write_then_read()` call by a `spi_sync()` one. The write `spi_transfer` now sets a 6 micro secs delay before the begin of the read transfer. This should be enough time for the conversion complete according to the datasheet. To test this, I tried to force doing `spi_sync`s with less than 6 usecs between each other but the less I got was a 20 usec interval between them (without the `delay_usecs` property set). With the `delay_usecs` set to 6, the interval was about 25 usecs between each `spi_sync` on average. Would you want me to do more tests? Would you want to suggest any test to be made?

marceloschmitt1 commented:
### Changelog V6:
- Renamed `AD7292_ADC_RSLT` macro to `AD7292_ADC_DATA`
- Stored the result of the conversion in a local variable to then use the ADC data macro
- Changed indentation at channel definition
- Defined masks to retrieve channel sampling mode and channel VIN range
- Added checks on returns of functions that rely on SPI transfers
- Minor changes regarding variable creation
- Created a separate function to retrieve a given channel input range multiplier
- Added comments on the `scale` calculation scheme
- Improved commit description

marceloschmitt1 commented:
### Changelog V7:
- set up one `spi_transfer` for writing with a 6 usec delay after it
- set up one `spi_transfer` for reading
- replaced `spi_write_then_read()` by `spi_sync()`

stefpopa commented:
This looks good from my side. It will be merged later if there are no objections.

marceloschmitt1 commented:
I recall someone saying that the commits would have to be squashed into two, one for the driver, other for the dt doc. May I do it and send you these two patches by email for a final review?

stefpopa commented:
Hi Marcelo,
Yes, something like this:
https://git.kernel.org/pub/scm/linux/kernel/git/jic23/iio.git/commit/?h=testing&id=a5f8c7da3dbee20b5acf1a8e24139956255aef57
https://git.kernel.org/pub/scm/linux/kernel/git/jic23/iio.git/commit/?h=testing&id=b64d97300888a51b0c122e96457faa7059b6987b
